lib/src/executable.dart

```diff
@@ -101,21 +112,103 @@ Future<void> handleFastFormat(List<String> args) async {
   }
 }
 
-void generateRunScript() {
-  if (shouldWriteRunScript) {
+List<String> generateRunScript() {
+  // Generate the run script if it doesn't yet exist or regenerate it if the
+  // existing script is outdated.
+  final runScriptContents = buildDartDevRunScriptContents();
+  if (!_runScript.existsSync() ||
+      _runScript.readAsStringSync() != runScriptContents) {
     logTimedSync(log, 'Generating run script', () {
       createCacheDir();
       _runScript.writeAsStringSync(buildDartDevRunScriptContents());
     }, level: Level.INFO);
   }
-}
 
-bool get shouldWriteRunScript =>
-    !_runScript.existsSync() ||
-    _runScript.readAsStringSync() != buildDartDevRunScriptContents();
+  // Generate a digest of inputs to the run script. We use this to determine
+  // whether we need to recompile the executable.
+  String encodedDigest;
+  logTimedSync(log, 'Computing run script digest', () {
+    var configHasRelativeImports = false;
+    var configHasSamePackageImports = false;
+    if (_config.existsSync()) {
+      final contents = _config.readAsStringSync();
+      configHasRelativeImports =
+          RegExp(r'''^import ['"][^:]+''').hasMatch(contents);
+      final currentPackageName =
+          Pubspec.parse(File('pubspec.yaml').readAsStringSync()).name;
+      configHasSamePackageImports =
+          RegExp('''import ['"]package:$currentPackageName''')
+              .hasMatch(contents);
+    }
+
+    if (configHasSamePackageImports) {
+      log.fine(
+          'Skipping compilation because $_configPath imports from its own package.');
+      // If the config imports from its own source files, we don't have a way of
+      // efficiently tracking changes that would require recompilation of this
+      // executable, so we skip the compilation altogether.
+      if (_runExecutable.existsSync()) {
+        _runExecutable.deleteSync();
+      }
+      if (_runExecutableDigest.existsSync()) {
+        _runExecutableDigest.deleteSync();
+      }
+      return;
+    }
+
+    final digest = md5.convert([
+      ..._packageConfig.readAsBytesSync(),
```

Comment:
I just thought about something... `path`-dependency packages can change without updating `package_config.dart`...

Depending on those within the `tool` directory should be fairly uncommon, but it could potentially cause issues if you're, say, developing a package that's being package-imported in the `tool/` directory, and are testing it out via a `path` dependency override...

I'm not sure how much we should worry about that case though, or what we'd be able to do about it without making this invalidation logic significantly more complex. 😕  What do you think?

Reply:
Good point.. I'm thinking it may be worth parsing the package_config.json and using it to lookup the type of dependency for each `package:` import in the dart_dev/config.dart file. If any of them are mutable, we skip the compilation & caching. This would also cover the same-package import scenario, which is nice because we're currently handling that separately.

Reply:
Ooh good idea!

I think you might need the pubspec.lock to look up the dependency type, though, since package_config.json only contains paths to the packages being depended on and not the dependency type.